## Imports

In [430]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Loading The Data

In [431]:
test = pd.read_csv("data/test.csv")
train = pd.read_csv("data/train.csv")

In [432]:
train.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


In [433]:
test.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name
0,0013_01,Earth,True,G/3/S,TRAPPIST-1e,27.0,False,0.0,0.0,0.0,0.0,0.0,Nelly Carsoning
1,0018_01,Earth,False,F/4/S,TRAPPIST-1e,19.0,False,0.0,9.0,0.0,2823.0,0.0,Lerome Peckers
2,0019_01,Europa,True,C/0/S,55 Cancri e,31.0,False,0.0,0.0,0.0,0.0,0.0,Sabih Unhearfus
3,0021_01,Europa,False,C/1/S,TRAPPIST-1e,38.0,False,0.0,6652.0,0.0,181.0,585.0,Meratz Caltilter
4,0023_01,Earth,False,F/5/S,TRAPPIST-1e,20.0,False,10.0,0.0,635.0,0.0,0.0,Brence Harperez


## Data Cleaning

In [434]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


In [435]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4277 entries, 0 to 4276
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   4277 non-null   object 
 1   HomePlanet    4190 non-null   object 
 2   CryoSleep     4184 non-null   object 
 3   Cabin         4177 non-null   object 
 4   Destination   4185 non-null   object 
 5   Age           4186 non-null   float64
 6   VIP           4184 non-null   object 
 7   RoomService   4195 non-null   float64
 8   FoodCourt     4171 non-null   float64
 9   ShoppingMall  4179 non-null   float64
 10  Spa           4176 non-null   float64
 11  VRDeck        4197 non-null   float64
 12  Name          4183 non-null   object 
dtypes: float64(6), object(7)
memory usage: 434.5+ KB


There are a total of 8693 observations in the training set while the test set has 4277 observations.

### Missing Values

In [436]:
train.isna().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [437]:
test.isna().sum()

PassengerId       0
HomePlanet       87
CryoSleep        93
Cabin           100
Destination      92
Age              91
VIP              93
RoomService      82
FoodCourt       106
ShoppingMall     98
Spa             101
VRDeck           80
Name             94
dtype: int64

In [438]:
train.isna().sum()/len(train)*100

PassengerId     0.000000
HomePlanet      2.312205
CryoSleep       2.496261
Cabin           2.289198
Destination     2.093639
Age             2.059128
VIP             2.335212
RoomService     2.082135
FoodCourt       2.105142
ShoppingMall    2.392730
Spa             2.105142
VRDeck          2.162660
Name            2.300702
Transported     0.000000
dtype: float64

In [439]:
test.isna().sum()/len(test)*100

PassengerId     0.000000
HomePlanet      2.034136
CryoSleep       2.174421
Cabin           2.338087
Destination     2.151040
Age             2.127660
VIP             2.174421
RoomService     1.917232
FoodCourt       2.478373
ShoppingMall    2.291326
Spa             2.361468
VRDeck          1.870470
Name            2.197802
dtype: float64

Over 2% missing values! Let's start filling them in one by one.

### HomePlanet

In [440]:
train["HomePlanet"].value_counts()

Earth     4602
Europa    2131
Mars      1759
Name: HomePlanet, dtype: int64

So, there are just three unique values in the `HomePlanet` column. Also, they are distributed quite evenly. We can't use the most common value to fill in the missing values. Let's see if they are related with some other column.

#### With `Desitination`

In [441]:
train[(train["HomePlanet"]=="Earth")]["Destination"].value_counts()/len(train[(train["HomePlanet"]=="Earth")])*100

TRAPPIST-1e      67.383746
PSO J318.5-22    15.471534
55 Cancri e      14.993481
Name: Destination, dtype: float64

In [442]:
train[(train["HomePlanet"]=="Europa")]["Destination"].value_counts()/len(train[(train["HomePlanet"]=="Europa")])*100

TRAPPIST-1e      55.795401
55 Cancri e      41.576725
PSO J318.5-22     0.891600
Name: Destination, dtype: float64

In [443]:
train[(train["HomePlanet"]=="Mars")]["Destination"].value_counts()/len(train[(train["HomePlanet"]=="Mars")])*100

TRAPPIST-1e      83.854463
55 Cancri e      10.972143
PSO J318.5-22     2.785674
Name: Destination, dtype: float64

#### With `Cabin`

In [444]:
train["Cabin"]

0          B/0/P
1          F/0/S
2          A/0/S
3          A/0/S
4          F/1/S
          ...   
8688      A/98/P
8689    G/1499/S
8690    G/1500/S
8691     E/608/S
8692     E/608/S
Name: Cabin, Length: 8693, dtype: object

In [445]:
train["CabinFirst"] = train["Cabin"].str[0]
train["CabinLast"] = train["Cabin"].str[-1]
test["CabinFirst"] = test["Cabin"].str[0]
test["CabinLast"] = test["Cabin"].str[-1]

In [446]:
train[(train["HomePlanet"]=="Earth")]["CabinFirst"].value_counts()/len(train[(train["HomePlanet"]=="Earth")])*100

G    54.280748
F    35.071708
E     8.583225
Name: CabinFirst, dtype: float64

In [447]:
train[(train["HomePlanet"]=="Earth")]["CabinLast"].value_counts()/len(train[(train["HomePlanet"]=="Earth")])*100

P    49.32638
S    48.60930
Name: CabinLast, dtype: float64

In [448]:
train[(train["HomePlanet"]=="Mars")]["CabinFirst"].value_counts()/len(train[(train["HomePlanet"]=="Mars")])*100

F    63.104036
E    18.760659
D    16.031836
Name: CabinFirst, dtype: float64

In [449]:
train[(train["HomePlanet"]=="Mars")]["CabinLast"].value_counts()/len(train[(train["HomePlanet"]=="Mars")])*100

P    50.483229
S    47.413303
Name: CabinLast, dtype: float64

In [450]:
train[(train["HomePlanet"]=="Europa")]["CabinFirst"].value_counts()/len(train[(train["HomePlanet"]=="Europa")])*100

B    35.945565
C    34.443923
A    11.825434
D     8.728297
E     6.006570
T     0.187705
Name: CabinFirst, dtype: float64

In [451]:
train[(train["HomePlanet"]=="Europa")]["CabinLast"].value_counts()/len(train[(train["HomePlanet"]=="Europa")])*100

S    52.322853
P    44.814641
Name: CabinLast, dtype: float64

There is not any simple pattern between `HomePlanet` and other columns. However, instead of filling the `HomePlanet` with random values, we can try a classifier to predict the missing values using other columns. 

#### Using Classifier to Fill NA values

##### Preparing the Data

In [452]:
train[train["HomePlanet"].isnull()].isna().sum()

PassengerId       0
HomePlanet      201
CryoSleep         2
Cabin             6
Destination       4
Age               2
VIP               3
RoomService       6
FoodCourt         2
ShoppingMall      7
Spa               3
VRDeck            1
Name              7
Transported       0
CabinFirst        6
CabinLast         6
dtype: int64

In [453]:
train[train["HomePlanet"].notna()].isna().sum()

PassengerId       0
HomePlanet        0
CryoSleep       215
Cabin           193
Destination     178
Age             177
VIP             200
RoomService     175
FoodCourt       181
ShoppingMall    201
Spa             180
VRDeck          187
Name            193
Transported       0
CabinFirst      193
CabinLast       193
dtype: int64

There are a couple of missing values in other columns. For the sake of simplicity, while training a classifier to predict the missing values, we'll fill them with the most common value or some other statistical significant value.

In [454]:
df = pd.concat([train, test])

In [455]:
for_homeplanet_train = df[df["HomePlanet"].notna()]
for_homeplanet_test = df[df["HomePlanet"].isna()]

Let's drop some of the `NA` values. Also drop the `PassengerId` column.

In [456]:
for_homeplanet_train = for_homeplanet_train.drop(["Transported"], axis=1)

In [457]:
for_homeplanet_test = for_homeplanet_test.drop(["Transported"], axis=1)

In [458]:
for_homeplanet_train.isna().sum()

PassengerId       0
HomePlanet        0
CryoSleep       305
Cabin           292
Destination     268
Age             265
VIP             293
RoomService     257
FoodCourt       285
ShoppingMall    296
Spa             275
VRDeck          265
Name            284
CabinFirst      292
CabinLast       292
dtype: int64

In [459]:
for_homeplanet_train[for_homeplanet_train["CryoSleep"].isna()].isna().sum()

PassengerId       0
HomePlanet        0
CryoSleep       305
Cabin            10
Destination       7
Age               3
VIP               9
RoomService       3
FoodCourt         3
ShoppingMall     10
Spa               7
VRDeck            8
Name             11
CabinFirst       10
CabinLast        10
dtype: int64

It won't do us any good if we drop `NA` valued row for a couple of columns. For example, if we drop all rows with `NA` in the `CryoSleep` column, other columns still have almost the same number of missing values. So, let's just fill `NA` with most common values.

In [460]:
columns_to_fill = ["CryoSleep", "Destination", "VIP", "CabinFirst", "CabinLast"]
for col in columns_to_fill:
    for_homeplanet_train[col].fillna(for_homeplanet_train[col].value_counts().index[0], inplace=True)
    for_homeplanet_test[col].fillna(for_homeplanet_test[col].value_counts().index[0], inplace=True)

In [461]:
for_homeplanet_train.isna().sum()

PassengerId       0
HomePlanet        0
CryoSleep         0
Cabin           292
Destination       0
Age             265
VIP               0
RoomService     257
FoodCourt       285
ShoppingMall    296
Spa             275
VRDeck          265
Name            284
CabinFirst        0
CabinLast         0
dtype: int64

In [462]:
for_homeplanet_test.isna().sum()

PassengerId       0
HomePlanet      288
CryoSleep         0
Cabin             7
Destination       0
Age               5
VIP               0
RoomService       6
FoodCourt         4
ShoppingMall     10
Spa               9
VRDeck            3
Name             10
CabinFirst        0
CabinLast         0
dtype: int64

Numerical columns are going to be filled with the mean of the column.

In [463]:
columns_to_fill = ["Age", "RoomService", "FoodCourt", "ShoppingMall", "Spa", "VRDeck"]
for col in columns_to_fill:
    for_homeplanet_train[col].fillna(for_homeplanet_train[col].mean(), inplace=True)
    for_homeplanet_test[col].fillna(for_homeplanet_test[col].mean(), inplace=True)
# for_homeplanet_train["Age"].fillna(for_homeplanet_train["Age"].mean(), inplace=True)

In [464]:
for_homeplanet_train.isna().sum()

PassengerId       0
HomePlanet        0
CryoSleep         0
Cabin           292
Destination       0
Age               0
VIP               0
RoomService       0
FoodCourt         0
ShoppingMall      0
Spa               0
VRDeck            0
Name            284
CabinFirst        0
CabinLast         0
dtype: int64

In [465]:
for_homeplanet_test.isna().sum()

PassengerId       0
HomePlanet      288
CryoSleep         0
Cabin             7
Destination       0
Age               0
VIP               0
RoomService       0
FoodCourt         0
ShoppingMall      0
Spa               0
VRDeck            0
Name             10
CabinFirst        0
CabinLast         0
dtype: int64

In [466]:
train_PI = for_homeplanet_train["PassengerId"]
train_name = for_homeplanet_train["Name"]
train_cabin = for_homeplanet_train["Cabin"]
test_PI = for_homeplanet_test["PassengerId"]
test_name = for_homeplanet_test["Name"]
test_cabin = for_homeplanet_test["Cabin"]
for_homeplanet_train.drop(["PassengerId", "Name", "Cabin"], axis=1, inplace=True)
for_homeplanet_test.drop(["PassengerId", "Name", "Cabin", "HomePlanet"], axis=1, inplace=True)

In [467]:
for_homeplanet_train.dtypes

HomePlanet       object
CryoSleep          bool
Destination      object
Age             float64
VIP                bool
RoomService     float64
FoodCourt       float64
ShoppingMall    float64
Spa             float64
VRDeck          float64
CabinFirst       object
CabinLast        object
dtype: object

In [468]:
for_homeplanet_train["CryoSleep"] = for_homeplanet_train["CryoSleep"].astype(int)
for_homeplanet_train["VIP"] = for_homeplanet_train["VIP"].astype(int)

for_homeplanet_test["CryoSleep"] = for_homeplanet_test["CryoSleep"].astype(int)
for_homeplanet_test["VIP"] = for_homeplanet_test["VIP"].astype(int)

In [469]:
X = for_homeplanet_train.drop(["HomePlanet"], axis=1)
y = for_homeplanet_train["HomePlanet"]

In [470]:
X = pd.get_dummies(X, drop_first=True)
homeplanet_test = pd.get_dummies(for_homeplanet_test, drop_first=True)

In [471]:
X.head()

,CryoSleep,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Destination_PSO J318.5-22,Destination_TRAPPIST-1e,CabinFirst_B,CabinFirst_C,CabinFirst_D,CabinFirst_E,CabinFirst_F,CabinFirst_G,CabinFirst_T,CabinLast_S
0,0,39.0,0,0.0,0.0,0.0,0.0,0.0,0,1,1,0,0,0,0,0,0,0
1,0,24.0,0,109.0,9.0,25.0,549.0,44.0,0,1,0,0,0,0,1,0,0,1
2,0,58.0,1,43.0,3576.0,0.0,6715.0,49.0,0,1,0,0,0,0,0,0,0,1
3,0,33.0,0,0.0,1283.0,371.0,3329.0,193.0,0,1,0,0,0,0,0,0,0,1
4,0,16.0,0,303.0,70.0,151.0,565.0,2.0,0,1,0,0,0,0,1,0,0,1


#### Training a Classifier

In [472]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [473]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [474]:
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)
X_test = scaler.fit_transform(X_test)

In [475]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
lr = LogisticRegression(max_iter=1000)
lr.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

In [476]:
accuracy_score(y_test, lr.predict(X_test))

0.892392589672842

In [477]:
accuracy_score(y_train, lr.predict(X_train))

0.8990635781172992

In [478]:
y_pred = lr.predict(homeplanet_test)


c:\Users\harik\anaconda3\envs\portfolio\lib\site-packages\sklearn\base.py:444: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  f"X has feature names, but {self.__class__.__name__} was fitted without"


In [479]:
homeplanet_test["HomePlanet"] = y_pred
homeplanet_test["Name"] = test_name
homeplanet_test["PassengerId"] = test_PI
homeplanet_test["Cabin"] = test_cabin

In [480]:
homeplanet_test

,CryoSleep,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Destination_PSO J318.5-22,Destination_TRAPPIST-1e,...,CabinFirst_D,CabinFirst_E,CabinFirst_F,CabinFirst_G,CabinFirst_T,CabinLast_S,HomePlanet,Name,PassengerId,Cabin
59,1,33.0,0,0.0,0.0,182.251799,0.0,0.0,0,1,...,0,1,0,0,0,1,Mars,Colatz Keen,0064_02,E/3/S
113,0,39.0,0,0.0,2344.0,0.000000,65.0,6898.0,0,1,...,0,0,0,0,0,0,Europa,Batan Coning,0119_01,A/0/P
186,1,24.0,0,0.0,0.0,182.251799,0.0,0.0,0,0,...,1,0,0,0,0,0,Mars,Arraid Inicont,0210_01,D/6/P
225,0,18.0,0,313.0,1.0,691.000000,283.0,0.0,0,1,...,0,0,1,0,0,1,Europa,Almone Sté,0242_01,F/46/S
234,1,54.0,0,0.0,0.0,0.000000,0.0,0.0,0,0,...,0,0,0,0,0,1,Europa,Diphah Amsive,0251_01,C/11/S
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3957,0,19.0,0,4.0,0.0,1604.000000,0.0,0.0,0,1,...,0,1,0,0,0,0,Mars,Vanley Simmonders,8621_01,E/552/P
3980,1,9.0,0,0.0,0.0,0.000000,0.0,0.0,0,0,...,0,0,0,1,0,1,Earth,Eilan Kellson,8678_01,G/1399/S
4032,1,40.0,0,0.0,0.0,0.000000,0.0,0.0,0,1,...,1,0,0,0,0,0,Europa,Raston Maltorted,8775_01,D/275/P
4159,0,42.0,0,0.0,0.0,28.000000,726.0,0.0,0,1,...,0,0,0,1,0,1,Europa,Ale Whitersone,9025_01,G/1454/S


In [481]:
X["HomePlanet"] = y
X["Name"] = train_name
X["PassengerId"] = train_PI
X["Cabin"] = train_cabin

In [482]:
pd.concat([X, homeplanet_test])["HomePlanet"]

0       Europa
1        Earth
2       Europa
3       Europa
4        Earth
         ...  
3957      Mars
3980     Earth
4032    Europa
4159    Europa
4213      Mars
Name: HomePlanet, Length: 12970, dtype: object

In [483]:
all = pd.concat([X, homeplanet_test])

In [484]:
all["PassengerId"]

0       0001_01
1       0002_01
2       0003_01
3       0003_02
4       0004_01
         ...   
3957    8621_01
3980    8678_01
4032    8775_01
4159    9025_01
4213    9146_01
Name: PassengerId, Length: 12970, dtype: object

In [485]:
train

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported,CabinFirst,CabinLast
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False,B,P
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True,F,S
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False,A,S
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False,A,S
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True,F,S
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
8688,9276_01,Europa,False,A/98/P,55 Cancri e,41.0,True,0.0,6819.0,0.0,1643.0,74.0,Gravior Noxnuther,False,A,P
8689,9278_01,Earth,True,G/1499/S,PSO J318.5-22,18.0,False,0.0,0.0,0.0,0.0,0.0,Kurta Mondalley,False,G,S
8690,9279_01,Earth,False,G/1500/S,TRAPPIST-1e,26.0,False,0.0,0.0,1872.0,1.0,0.0,Fayey Connon,True,G,S
8691,9280_01,Europa,False,E/608/S,55 Cancri e,32.0,False,0.0,1049.0,0.0,353.0,3235.0,Celeon Hontichre,False,E,S


In [489]:
train = pd.merge(train, all[["PassengerId", "HomePlanet"]], on="PassengerId")
test = pd.merge(test, all[["PassengerId", "HomePlanet"]], on="PassengerId")

In [490]:
train.isna().sum()

PassengerId       0
HomePlanet_x    201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
CabinFirst      199
CabinLast       199
HomePlanet_y      0
dtype: int64

In [491]:
train.drop("HomePlanet_x", axis=1, inplace=True)
test.drop("HomePlanet_x", axis=1, inplace=True)

In [492]:
train.to_csv("data/train_with_homeplanet.csv", index=False)
test.to_csv("data/test_with_homeplanet.csv", index=False)